File: src/chacana/checker.py
"""Static type checker for Chacana AST.

Rules 1 (contraction), 2 (free index invariance), 3 (symmetry validity).
"""

from __future__ import annotations

from collections import Counter
from collections.abc import Callable

from chacana.ast import (
    HEAD_ADD,
    HEAD_DETERMINANT,
    HEAD_DIVIDE,
    HEAD_HODGE_STAR,
    HEAD_INTERIOR_PRODUCT,
    HEAD_INVERSE,
    HEAD_LIE_DERIVATIVE,
    HEAD_MULTIPLY,
    HEAD_NEGATE,
    HEAD_NUMBER,
    HEAD_RATIONAL,
    HEAD_TRACE,
    HEAD_WEDGE,
    ChacanaIndex,
    ValidationToken,
    Variance,
    walk_tokens,
)
from chacana.context import GlobalContext
from chacana.errors import ChacanaTypeError
from chacana.indices import IndexAnalyzer


def _check_contraction(token: ValidationToken, ctx: GlobalContext | None) -> None:
    """Rule 1: Contraction indices must have opposite variance and matching index_type."""
    analyzer = IndexAnalyzer(ctx)
    _check_contraction_with(token, ctx, analyzer)


def _check_contraction_with(
    token: ValidationToken, ctx: GlobalContext | None, analyzer: IndexAnalyzer
) -> None:
    if token.head in (HEAD_MULTIPLY, HEAD_WEDGE, HEAD_DIVIDE):
        all_indices: list[ChacanaIndex] = []
        for arg in token.args:
            _check_contraction_with(arg, ctx, analyzer)
            all_indices.extend(analyzer.all_indices(arg))

        by_label: dict[str, list[ChacanaIndex]] = {}
        for idx in all_indices:
            by_label.setdefault(idx.label, []).append(idx)
        for label, group in by_label.items():
            if len(group) == 2:
                if group[0].index_type != group[1].index_type:
                    raise ChacanaTypeError(
                        f"Contraction index '{label}' has mismatched index type: "
                        f"{group[0].index_type.value} vs {group[1].index_type.value}"
                    )
                # Variance check for Multiply/Divide (wedge doesn't contract).
                is_product = token.head in (HEAD_MULTIPLY, HEAD_DIVIDE)
                if is_product and group[0].variance == group[1].variance:
                    if ctx and ctx.active_metric:
                        continue
                    raise ChacanaTypeError(
                        f"Contraction index '{label}' appears twice with same "
                        f"variance ({group[0].variance.value})"
                    )
            elif len(group) > 2:
                raise ChacanaTypeError(
                    f"Index '{label}' appears {len(group)} times (expected at most 2)"
                )
    else:
        for arg in token.args:
            _check_contraction_with(arg, ctx, analyzer)


def _check_free_index_invariance(token: ValidationToken, ctx: GlobalContext | None) -> None:
    """Rule 2: All terms in a sum must have the same free indices."""
    analyzer = IndexAnalyzer(ctx)
    for t in walk_tokens(token):
        if t.head != HEAD_ADD or len(t.args) < 2:
            continue
        ref = analyzer.free_indices(t.args[0])
        ref_counted = Counter((idx.label, idx.variance) for idx in ref)
        for i, arg in enumerate(t.args[1:], 1):
            arg_free = analyzer.free_indices(arg)
            arg_counted = Counter((idx.label, idx.variance) for idx in arg_free)
            if ref_counted != arg_counted:
                raise ChacanaTypeError(
                    f"Free index mismatch in sum: term 0 has "
                    f"{_format_indices(ref)}, term {i} has "
                    f"{_format_indices(arg_free)}"
                )


def _check_symmetry(token: ValidationToken, ctx: GlobalContext | None) -> None:
    """Rule 3: Symmetrized index groups must have matching variance and index_type.

    Checks two sources of symmetry information:
    1. Explicit symmetrization in the expression
       (metadata: symmetrized_groups / antisymmetrized_groups)
    2. Declared symmetries in the context (tensor.symmetries)
    """
    for t in walk_tokens(token):
        _check_symmetry_single(t, ctx)


def _validate_symmetry_group(
    indices: list[ChacanaIndex], positions: list[int], label: str
) -> None:
    """Check that indices at the given positions have matching variance and type."""
    if len(positions) < 2:
        return
    ref = indices[positions[0]]
    for pos in positions[1:]:
        other = indices[pos]
        if ref.variance != other.variance:
            raise ChacanaTypeError(
                f"Variance mismatch in {label}: index "
                f"'{ref.label}' ({ref.variance.value}) vs "
                f"'{other.label}' ({other.variance.value})"
            )
        if ref.index_type != other.index_type:
            raise ChacanaTypeError(
                f"Index type mismatch in {label}: index "
                f"'{ref.label}' ({ref.index_type.value}) vs "
                f"'{other.label}' ({other.index_type.value})"
            )


def _check_symmetry_single(token: ValidationToken, ctx: GlobalContext | None) -> None:
    for kind, groups in (
        ("symmetrization", token.metadata.symmetrized_groups),
        ("anti-symmetrization", token.metadata.antisymmetrized_groups),
    ):
        for group in groups:
            _validate_symmetry_group(token.indices, group, kind)

    if ctx is not None:
        decl = ctx.tensors.get(token.head)
        if decl is not None and token.indices:
            for sym in decl.symmetries:
                positions = [i - 1 for i in sym.indices]
                if all(0 <= p < len(token.indices) for p in positions):
                    _validate_symmetry_group(
                        token.indices,
                        positions,
                        f"declared symmetry of '{token.head}'",
                    )


_STRUCTURAL_HEADS = frozenset(
    {
        HEAD_ADD,
        HEAD_MULTIPLY,
        HEAD_WEDGE,
        HEAD_NEGATE,
        HEAD_NUMBER,
        HEAD_DIVIDE,
        HEAD_RATIONAL,
    }
)


def _check_rank(token: ValidationToken, ctx: GlobalContext) -> None:
    """Check that tensor usage matches declared rank and index pattern."""
    for t in walk_tokens(token):
        if t.head in _STRUCTURAL_HEADS:
            continue
        decl = ctx.tensors.get(t.head)
        if decl is None:
            continue
        # Derivative indices (;e, ,a) are not part of the tensor's intrinsic rank
        tensor_indices = [idx for idx in t.indices if not idx.is_derivative]
        if tensor_indices and len(tensor_indices) != decl.rank:
            diff = decl.rank - len(tensor_indices)
            if not (ctx.active_metric and diff > 0 and diff % 2 == 0):
                raise ChacanaTypeError(
                    f"Tensor '{t.head}' declared with rank {decl.rank}, "
                    f"but used with {len(tensor_indices)} indices"
                )
        if tensor_indices and decl.index_pattern and not ctx.active_metric:
            for i, (actual, expected) in enumerate(
                zip(tensor_indices, decl.index_pattern, strict=False)
            ):
                if actual.variance != expected:
                    raise ChacanaTypeError(
                        f"Tensor '{t.head}' index {i}: expected "
                        f"{expected.value}, got {actual.variance.value}"
                    )


def _resolve_rank(token: ValidationToken, ctx: GlobalContext) -> int | None:
    """Resolve the rank of a token from its indices or context declaration."""
    if token.indices:
        return len(token.indices)
    decl = ctx.tensors.get(token.head)
    if decl is not None:
        return decl.rank
    return None


def _resolve_index_pattern(token: ValidationToken, ctx: GlobalContext) -> list[Variance] | None:
    """Resolve the index pattern (variance list) for a token."""
    if token.indices:
        return [idx.variance for idx in token.indices]
    decl = ctx.tensors.get(token.head)
    if decl is not None and decl.index_pattern:
        return list(decl.index_pattern)
    return None


def _is_vector(token: ValidationToken, ctx: GlobalContext) -> bool | None:
    """Check if a token represents a vector field (rank 1, contravariant).

    Returns True/False if determinable, None if unknown.
    """
    rank = _resolve_rank(token, ctx)
    if rank is None:
        return None
    if rank != 1:
        return False
    pattern = _resolve_index_pattern(token, ctx)
    if pattern is None:
        return None
    return pattern[0] == Variance.CONTRA


OperatorCheck = Callable[[ValidationToken, GlobalContext], None]


def _requires_metric(token: ValidationToken, ctx: GlobalContext) -> None:
    if not ctx.active_metric:
        raise ChacanaTypeError("Hodge star operator requires an active_metric in the context")


def _first_arg_is_vector(op_name: str) -> OperatorCheck:
    def check(token: ValidationToken, ctx: GlobalContext) -> None:
        if len(token.args) < 1:
            return
        if _is_vector(token.args[0], ctx) is False:
            raise ChacanaTypeError(
                f"{op_name} first argument must be a vector field (rank 1 contravariant)"
            )

    return check


def _second_arg_not_zero_form(token: ValidationToken, ctx: GlobalContext) -> None:
    if len(token.args) < 2:
        return
    rank = _resolve_rank(token.args[1], ctx)
    if rank is not None and rank == 0:
        raise ChacanaTypeError(
            "Interior product is undefined for 0-forms (rank 0); "
            "second argument must be a p-form with p >= 1"
        )


def _first_arg_min_rank(min_rank: int, op_name: str) -> OperatorCheck:
    def check(token: ValidationToken, ctx: GlobalContext) -> None:
        if len(token.args) < 1:
            return
        rank = _resolve_rank(token.args[0], ctx)
        if rank is not None and rank < min_rank:
            raise ChacanaTypeError(
                f"{op_name} requires a tensor of rank >= {min_rank}, but argument has rank {rank}"
            )

    return check


def _first_arg_exact_rank(exact_rank: int, op_name: str) -> OperatorCheck:
    def check(token: ValidationToken, ctx: GlobalContext) -> None:
        if len(token.args) < 1:
            return
        rank = _resolve_rank(token.args[0], ctx)
        if rank is not None and rank != exact_rank:
            raise ChacanaTypeError(
                f"{op_name} requires a rank-{exact_rank} tensor, but argument has rank {rank}"
            )

    return check


_OPERATOR_CONSTRAINTS: dict[str, list[OperatorCheck]] = {
    HEAD_HODGE_STAR: [_requires_metric],
    HEAD_INTERIOR_PRODUCT: [
        _first_arg_is_vector("Interior product"),
        _second_arg_not_zero_form,
    ],
    HEAD_LIE_DERIVATIVE: [_first_arg_is_vector("Lie derivative")],
    HEAD_TRACE: [_first_arg_min_rank(2, "Trace")],
    HEAD_DETERMINANT: [_first_arg_exact_rank(2, "Determinant")],
    HEAD_INVERSE: [_first_arg_exact_rank(2, "Inverse")],
}


def _check_operators(token: ValidationToken, ctx: GlobalContext) -> None:
    """Check differential geometry operator constraints."""
    for t in walk_tokens(token):
        for constraint in _OPERATOR_CONSTRAINTS.get(t.head, []):
            constraint(t, ctx)


def _format_indices(indices: list[ChacanaIndex]) -> str:
    parts = []
    for idx in indices:
        prefix = "^" if idx.variance == Variance.CONTRA else "_"
        parts.append(f"{prefix}{idx.label}")
    return "{" + " ".join(parts) + "}" if parts else "{}"


def check(token: ValidationToken, ctx: GlobalContext | None = None) -> ValidationToken:
    """Run all type checks on a ValidationToken. Returns the token if valid."""
    _check_contraction(token, ctx)
    _check_free_index_invariance(token, ctx)
    _check_symmetry(token, ctx)
    if ctx is not None:
        _check_rank(token, ctx)
        _check_operators(token, ctx)
    return token

File: vscode-chacana/src/server/checker.ts
/**
 * Static type checker for Chacana AST.
 * Ports src/chacana/checker.py to TypeScript.
 *
 * Key difference from Python: collects ALL errors instead of throwing
 * on the first one, which is the expected LSP behavior.
 */

import {
  type ChacanaIndex,
  type ValidationToken,
  type SourceRange,
  Variance,
  HEAD_ADD,
  HEAD_NEGATE,
  HEAD_MULTIPLY,
  HEAD_WEDGE,
  HEAD_EXTERIOR_DERIVATIVE,
  HEAD_LIE_DERIVATIVE,
  HEAD_TRACE,
  HEAD_DETERMINANT,
  HEAD_INVERSE,
  HEAD_HODGE_STAR,
  HEAD_INTERIOR_PRODUCT,
  HEAD_NUMBER,
  walkTokens,
} from "./ast.js";
import type { GlobalContext } from "./context.js";

export interface CheckerDiagnostic {
  message: string;
  range: SourceRange | null;
  code: string;
}

// ── Index analysis helpers (metric-aware) ─────────────────────────

function freeIndices(token: ValidationToken, metricAware: boolean = false): ChacanaIndex[] {
  if (token.head === HEAD_ADD) {
    return token.args.length > 0 ? freeIndices(token.args[0], metricAware) : [];
  }
  if (token.head === HEAD_NEGATE) {
    return token.args.length > 0 ? freeIndices(token.args[0], metricAware) : [];
  }
  if (token.head === HEAD_MULTIPLY) {
    const all: ChacanaIndex[] = [];
    for (const arg of token.args) all.push(...freeIndices(arg, metricAware));
    return removeContracted(all, metricAware);
  }
  if (token.head === HEAD_WEDGE) {
    const all: ChacanaIndex[] = [];
    for (const arg of token.args) all.push(...freeIndices(arg, metricAware));
    return all;
  }
  if (token.head === HEAD_EXTERIOR_DERIVATIVE) {
    return token.args.length > 0 ? freeIndices(token.args[0], metricAware) : [];
  }
  if (token.head === HEAD_NUMBER) return [];
  return [...token.indices];
}

function removeContracted(indices: ChacanaIndex[], metricAware: boolean = false): ChacanaIndex[] {
  const free: ChacanaIndex[] = [];
  const consumed = new Set<number>();
  for (let i = 0; i < indices.length; i++) {
    if (consumed.has(i)) continue;
    let foundPair = false;
    for (let j = i + 1; j < indices.length; j++) {
      if (consumed.has(j)) continue;
      if (
        indices[i].label === indices[j].label &&
        indices[i].indexType === indices[j].indexType &&
        (indices[i].variance !== indices[j].variance || metricAware)
      ) {
        consumed.add(i);
        consumed.add(j);
        foundPair = true;
        break;
      }
    }
    if (!foundPair) free.push(indices[i]);
  }
  return free;
}

function getAllIndices(token: ValidationToken): ChacanaIndex[] {
  if (token.head === HEAD_ADD) {
    if (token.args.length === 0) return [];
    // Max-multiplicity union across all terms (fixes y74).
    const maxCounts = new Map<string, { idx: ChacanaIndex; count: number }>();
    for (const arg of token.args) {
      const argCounts = new Map<string, { idx: ChacanaIndex; count: number }>();
      for (const idx of getAllIndices(arg)) {
        const key = `${idx.label}:${idx.variance}:${idx.indexType}`;
        const entry = argCounts.get(key);
        if (entry) entry.count++;
        else argCounts.set(key, { idx, count: 1 });
      }
      for (const [key, { idx, count }] of argCounts) {
        const existing = maxCounts.get(key);
        if (!existing || count > existing.count) {
          maxCounts.set(key, { idx, count });
        }
      }
    }
    const result: ChacanaIndex[] = [];
    for (const { idx, count } of maxCounts.values()) {
      for (let i = 0; i < count; i++) result.push(idx);
    }
    return result;
  }
  if (token.head === HEAD_NEGATE) {
    return token.args.length > 0 ? getAllIndices(token.args[0]) : [];
  }
  if (token.head === HEAD_MULTIPLY || token.head === HEAD_WEDGE) {
    const result: ChacanaIndex[] = [];
    for (const arg of token.args) result.push(...getAllIndices(arg));
    return result;
  }
  if (token.head === HEAD_NUMBER) return [];
  return [...token.indices];
}

function resolveRank(
  token: ValidationToken,
  ctx: GlobalContext,
): number | null {
  if (token.indices.length > 0) return token.indices.length;
  const decl = ctx.tensors.get(token.head);
  return decl?.rank ?? null;
}

function resolveIndexPattern(
  token: ValidationToken,
  ctx: GlobalContext,
): Variance[] | null {
  if (token.indices.length > 0) return token.indices.map((i) => i.variance);
  const decl = ctx.tensors.get(token.head);
  return decl?.indexPattern.length ? [...decl.indexPattern] : null;
}

function isVector(
  token: ValidationToken,
  ctx: GlobalContext,
): boolean | null {
  const rank = resolveRank(token, ctx);
  if (rank == null) return null;
  if (rank !== 1) return false;
  const pattern = resolveIndexPattern(token, ctx);
  if (pattern == null) return null;
  return pattern[0] === Variance.Contra;
}

function formatIndices(indices: ChacanaIndex[]): string {
  if (indices.length === 0) return "{}";
  const parts = indices.map(
    (i) => `${i.variance === Variance.Contra ? "^" : "_"}${i.label}`,
  );
  return `{${parts.join(" ")}}`;
}

// ── Rule 1: Contraction ────────────────────────────────────────────

function checkContraction(
  token: ValidationToken,
  ctx: GlobalContext | null,
  diags: CheckerDiagnostic[],
): void {
  if (token.head === HEAD_MULTIPLY || token.head === HEAD_WEDGE) {
    const allIdx: ChacanaIndex[] = [];
    for (const arg of token.args) {
      checkContraction(arg, ctx, diags);
      allIdx.push(...getAllIndices(arg));
    }

    const byLabel = new Map<string, ChacanaIndex[]>();
    for (const idx of allIdx) {
      const group = byLabel.get(idx.label) ?? [];
      group.push(idx);
      byLabel.set(idx.label, group);
    }
    for (const [label, group] of byLabel) {
      if (group.length === 2) {
        if (group[0].indexType !== group[1].indexType) {
          diags.push({
            message: `Contraction index '${label}' has mismatched index type: ${group[0].indexType} vs ${group[1].indexType}`,
            range: token.range,
            code: "chacana/contraction",
          });
        } else if (
          token.head === HEAD_MULTIPLY &&
          group[0].variance === group[1].variance
        ) {
          if (ctx?.activeMetric) continue;
          diags.push({
            message: `Contraction index '${label}' appears twice with same variance (${group[0].variance})`,
            range: token.range,
            code: "chacana/contraction",
          });
        }
      } else if (group.length > 2) {
        diags.push({
          message: `Index '${label}' appears ${group.length} times (expected at most 2)`,
          range: token.range,
          code: "chacana/contraction",
        });
      }
    }
  } else {
    for (const arg of token.args) checkContraction(arg, ctx, diags);
  }
}

// ── Rule 2: Free Index Invariance ──────────────────────────────────

function checkFreeIndexInvariance(
  token: ValidationToken,
  ctx: GlobalContext | null,
  diags: CheckerDiagnostic[],
): void {
  const metricAware = !!(ctx?.activeMetric);
  for (const t of walkTokens(token)) {
    if (t.head !== HEAD_ADD || t.args.length < 2) continue;
    const ref = freeIndices(t.args[0], metricAware);
    const refCounted = countIndices(ref);
    for (let i = 1; i < t.args.length; i++) {
      const argFree = freeIndices(t.args[i], metricAware);
      const argCounted = countIndices(argFree);
      if (!mapsEqual(refCounted, argCounted)) {
        diags.push({
          message: `Free index mismatch in sum: term 0 has ${formatIndices(ref)}, term ${i} has ${formatIndices(argFree)}`,
          range: t.range,
          code: "chacana/free-index",
        });
      }
    }
  }
}

function countIndices(indices: ChacanaIndex[]): Map<string, number> {
  const m = new Map<string, number>();
  for (const idx of indices) {
    const key = `${idx.label}:${idx.variance}`;
    m.set(key, (m.get(key) ?? 0) + 1);
  }
  return m;
}

function mapsEqual(a: Map<string, number>, b: Map<string, number>): boolean {
  if (a.size !== b.size) return false;
  for (const [k, v] of a) {
    if (b.get(k) !== v) return false;
  }
  return true;
}

// ── Rule 3: Symmetry ───────────────────────────────────────────────

function checkSymmetry(
  token: ValidationToken,
  ctx: GlobalContext | null,
  diags: CheckerDiagnostic[],
): void {
  for (const t of walkTokens(token)) {
    checkSymmetrySingle(t, ctx, diags);
  }
}

function validateSymmetryGroup(
  indices: ChacanaIndex[],
  positions: number[],
  label: string,
  range: SourceRange | null,
  diags: CheckerDiagnostic[],
): void {
  if (positions.length < 2) return;
  const ref = indices[positions[0]];
  for (let p = 1; p < positions.length; p++) {
    const other = indices[positions[p]];
    if (ref.variance !== other.variance) {
      diags.push({
        message: `Variance mismatch in ${label}: index '${ref.label}' (${ref.variance}) vs '${other.label}' (${other.variance})`,
        range,
        code: "chacana/symmetry",
      });
    }
    if (ref.indexType !== other.indexType) {
      diags.push({
        message: `Index type mismatch in ${label}: index '${ref.label}' (${ref.indexType}) vs '${other.label}' (${other.indexType})`,
        range,
        code: "chacana/symmetry",
      });
    }
  }
}

function checkSymmetrySingle(
  token: ValidationToken,
  ctx: GlobalContext | null,
  diags: CheckerDiagnostic[],
): void {
  for (const [kind, groups] of [
    ["symmetrization", token.metadata.symmetrized_groups],
    ["anti-symmetrization", token.metadata.antisymmetrized_groups],
  ] as const) {
    for (const group of groups) {
      validateSymmetryGroup(token.indices, group, kind, token.range, diags);
    }
  }

  if (ctx != null) {
    const decl = ctx.tensors.get(token.head);
    if (decl && token.indices.length > 0) {
      for (const sym of decl.symmetries) {
        const positions = sym.indices.map((i) => i - 1);
        if (positions.every((p) => p >= 0 && p < token.indices.length)) {
          validateSymmetryGroup(
            token.indices,
            positions,
            `declared symmetry of '${token.head}'`,
            token.range,
            diags,
          );
        }
      }
    }
  }
}

// ── Rule 4: Rank ───────────────────────────────────────────────────

const STRUCTURAL_HEADS = new Set([
  HEAD_ADD, HEAD_MULTIPLY, HEAD_WEDGE, HEAD_NEGATE, HEAD_NUMBER,
]);

function checkRank(
  token: ValidationToken,
  ctx: GlobalContext,
  diags: CheckerDiagnostic[],
): void {
  for (const t of walkTokens(token)) {
    if (STRUCTURAL_HEADS.has(t.head)) continue;
    const decl = ctx.tensors.get(t.head);
    if (!decl) continue;

    // Derivative indices (;e, ,a) are not part of the tensor's intrinsic rank
    const tensorIndices = t.indices.filter(idx => !idx.isDerivative);

    if (tensorIndices.length > 0 && tensorIndices.length !== decl.rank) {
      diags.push({
        message: `Tensor '${t.head}' declared with rank ${decl.rank}, but used with ${tensorIndices.length} indices`,
        range: t.range,
        code: "chacana/rank",
      });
    }

    if (tensorIndices.length > 0 && decl.indexPattern.length > 0 && !ctx.activeMetric) {
      for (let i = 0; i < Math.min(tensorIndices.length, decl.indexPattern.length); i++) {
        if (tensorIndices[i].variance !== decl.indexPattern[i]) {
          diags.push({
            message: `Tensor '${t.head}' index ${i}: expected ${decl.indexPattern[i]}, got ${tensorIndices[i].variance}`,
            range: t.range,
            code: "chacana/rank",
          });
        }
      }
    }
  }
}

// ── Rule 5: Operator Constraints ───────────────────────────────────

type OperatorCheck = (
  token: ValidationToken,
  ctx: GlobalContext,
  diags: CheckerDiagnostic[],
) => void;

function requiresMetric(
  token: ValidationToken,
  ctx: GlobalContext,
  diags: CheckerDiagnostic[],
): void {
  if (!ctx.activeMetric) {
    diags.push({
      message: "Hodge star operator requires an active_metric in the context",
      range: token.range,
      code: "chacana/operator",
    });
  }
}

function firstArgIsVector(opName: string): OperatorCheck {
  return (token, ctx, diags) => {
    if (token.args.length < 1) return;
    if (isVector(token.args[0], ctx) === false) {
      diags.push({
        message: `${opName} first argument must be a vector field (rank 1 contravariant)`,
        range: token.range,
        code: "chacana/operator",
      });
    }
  };
}

function secondArgNotZeroForm(
  token: ValidationToken,
  ctx: GlobalContext,
  diags: CheckerDiagnostic[],
): void {
  if (token.args.length < 2) return;
  const rank = resolveRank(token.args[1], ctx);
  if (rank != null && rank === 0) {
    diags.push({
      message: "Interior product is undefined for 0-forms (rank 0)",
      range: token.range,
      code: "chacana/operator",
    });
  }
}

function firstArgMinRank(minRank: number, opName: string): OperatorCheck {
  return (token, ctx, diags) => {
    if (token.args.length < 1) return;
    const rank = resolveRank(token.args[0], ctx);
    if (rank != null && rank < minRank) {
      diags.push({
        message: `${opName} requires a tensor of rank >= ${minRank}, but argument has rank ${rank}`,
        range: token.range,
        code: "chacana/operator",
      });
    }
  };
}

function firstArgExactRank(exactRank: number, opName: string): OperatorCheck {
  return (token, ctx, diags) => {
    if (token.args.length < 1) return;
    const rank = resolveRank(token.args[0], ctx);
    if (rank != null && rank !== exactRank) {
      diags.push({
        message: `${opName} requires a rank-${exactRank} tensor, but argument has rank ${rank}`,
        range: token.range,
        code: "chacana/operator",
      });
    }
  };
}

const OPERATOR_CONSTRAINTS: Record<string, OperatorCheck[]> = {
  [HEAD_HODGE_STAR]: [requiresMetric],
  [HEAD_INTERIOR_PRODUCT]: [firstArgIsVector("Interior product"), secondArgNotZeroForm],
  [HEAD_LIE_DERIVATIVE]: [firstArgIsVector("Lie derivative")],
  [HEAD_TRACE]: [firstArgMinRank(2, "Trace")],
  [HEAD_DETERMINANT]: [firstArgExactRank(2, "Determinant")],
  [HEAD_INVERSE]: [firstArgExactRank(2, "Inverse")],
};

function checkOperators(
  token: ValidationToken,
  ctx: GlobalContext,
  diags: CheckerDiagnostic[],
): void {
  for (const t of walkTokens(token)) {
    const constraints = OPERATOR_CONSTRAINTS[t.head];
    if (constraints) {
      for (const check of constraints) {
        check(t, ctx, diags);
      }
    }
  }
}

// ── Public API ─────────────────────────────────────────────────────

/**
 * Run all 5 checker rules. Returns an array of diagnostics (empty if valid).
 */
export function checkAll(
  token: ValidationToken,
  ctx: GlobalContext | null = null,
): CheckerDiagnostic[] {
  const diags: CheckerDiagnostic[] = [];
  checkContraction(token, ctx, diags);
  checkFreeIndexInvariance(token, ctx, diags);
  checkSymmetry(token, ctx, diags);
  if (ctx) {
    checkRank(token, ctx, diags);
    checkOperators(token, ctx, diags);
  }
  return diags;
}
